src/sentry/api/endpoints/project_codeowners.py

```diff
@@ -1,11 +1,156 @@
+            external_name__in=usernames,
+            organizationmember__organization=self.context["project"].organization,
+        )
+
+        external_users_names = [user.external_name for user in external_users]
+        external_users_diff = [name for name in usernames if name not in external_users_names]
+
+        if len(external_users_diff):
+            external_association_err.append(
+                f'The following usernames do not have an association in Sentry: {", ".join(external_users_diff)}.'
+            )
+
+        # Check if the team names have an association
+        external_teams = ExternalTeam.objects.filter(
+            external_name__in=teamnames,
+            team__organization=self.context["project"].organization,
+        )
+
+        external_teams_names = [team.external_name for team in external_teams]
+        external_teams_diff = [name for name in teamnames if name not in external_teams_names]
 
+        if len(external_teams_diff):
+            external_association_err.append(
+                f'The following team names do not have an association in Sentry: {", ".join(external_teams_diff)}.'
+            )
 
-class ProjectCodeOwnersEndpoint(ProjectEndpoint):
+        if len(external_association_err):
+            raise serializers.ValidationError({"raw": "\n".join(external_association_err)})
+
+        # Convert CODEOWNERS into IssueOwner syntax
+        users_dict = {
+            user.external_name: user.organizationmember.user.email for user in external_users
+        }
+        teams_dict = {team.external_name: f"#{team.team.slug}" for team in external_teams}
+        emails_dict = {email: email for email in emails}
+        associations = {**users_dict, **teams_dict, **emails_dict}
+
+        issue_owner_rules = convert_codeowners_syntax(
+            attrs["raw"], associations, attrs["code_mapping_id"]
+        )
+
+        # Convert IssueOwner syntax into schema syntax
+        validated_data = ProjectOwnershipSerializer(context=self.context).validate(
+            {"raw": issue_owner_rules}
+        )
+
+        return {**validated_data, **attrs}
+
+    def validate_code_mapping_id(self, code_mapping_id):
+        try:
+            return RepositoryProjectPathConfig.objects.get(
+                id=code_mapping_id, project=self.context["project"]
+            )
+        except RepositoryProjectPathConfig.DoesNotExist:
+            raise serializers.ValidationError("This code mapping does not exist.")
+
+    def create(self, validated_data):
+        # Create a project_ownership record with default values if none exists.
+        ownership = self.context["ownership"]
+        if ownership.id is None:
+            now = timezone.now()
+            ownership.date_created = now
+            ownership.last_updated = now
+            ownership.save()
+
+        # Save projectcodeowners record
+        repository_project_path_config = validated_data.pop("code_mapping_id", None)
```

Comment:
Do we expect the `code_mapping_id` not to be there? The `ProjectCodeOwners` isn't going to save without it so should we be raising an error?

Reply:
@MeredithAnya hmm, I don't expect "code_mapping_id" to never be there. It would have been caught earlier in the validation flow. Yes, also Postgres will throw a constraint error if we try to save without it. 

The None is just an explicit default to prevent a KeyError. Is this not pythonic?

Reply:
I guess my question is, why prevent a `KeyError` there just to get an `IntegrityError` later?

Reply:
hmm good question! To preface this discussion, we will never get to this point bc if `code_mapping_id` is missing, it would be caught by the validation flow.

`KeyError` is application-level and `IntegrityError` is db-level. I think there is more detail/context provided by an IntegrityError than KeyError. KeyError says that the key does not exist in the dictionary and thats it. But IntegrityError would provide us with  details on which column it failed to save on and why.